cpp/calculate-demand.cpp

```diff
@@ -72,68 +73,88 @@ long readUnsatisfiedDemand(std::string targetFilename) {
-  int xSize = targetDataset->GetRasterXSize();
-  int ySize = targetDataset->GetRasterYSize();
-  int nXBlocks = (xSize + xBlockSize - 1)/xBlockSize;
-  int nYBlocks = (ySize + yBlockSize - 1)/yBlockSize;
+
+  float* buffer = (float*) CPLMalloc(sizeof(float)*xBlockSize*yBlockSize);
+  BYTE* facilityBuffer = (BYTE*) CPLMalloc(sizeof(BYTE)*xBlockSize*yBlockSize);
+
+  float nodata = targetBand->GetNoDataValue();
+
+  int targetXSize = targetDataset->GetRasterXSize();
+  int targetYSize = targetDataset->GetRasterYSize();
+  int targetNXBlocks = (targetXSize + xBlockSize - 1)/xBlockSize;
+  int targetNYBlocks = (targetYSize + yBlockSize - 1)/yBlockSize;
   int nXValid, nYValid, xOffset, yOffset;
 
 #ifdef DEBUG
   std::cerr << "Target raster properties:" << std::endl;
-  std::cerr << " xSize " << xSize << std::endl;
-  std::cerr << " ySize " << ySize << std::endl;
+  std::cerr << " xSize " << targetXSize << std::endl;
+  std::cerr << " ySize " << targetYSize << std::endl;
   std::cerr << " xBlockSize " << xBlockSize << std::endl;
   std::cerr << " yBlockSize " << yBlockSize << std::endl;
-  std::cerr << " nYBlocks " << nYBlocks << std::endl;
-  std::cerr << " nXBlocks " << nXBlocks << std::endl;
+  std::cerr << " nYBlocks " << targetNYBlocks << std::endl;
+  std::cerr << " nXBlocks " << targetNXBlocks << std::endl;
 #endif
 
-  float* buffer = (float*) CPLMalloc(sizeof(float)*xBlockSize*yBlockSize);
-  BYTE* facilityBuffer = (BYTE*) CPLMalloc(sizeof(BYTE)*xBlockSize*yBlockSize);
-
-  float nodata = targetBand->GetNoDataValue();
+  for (size_t iFacility = 0; iFacility < facilities.size(); iFacility++) {
 
-  for (int iFacility = 0; iFacility < (int)facilities.size(); iFacility++) {
 #ifdef DEBUG
     std::cerr << "Processing facility " << facilities[iFacility] << std::endl;
 #endif
+
+    // open isochrone dataset
     GDALDataset* facilityDataset = openRaster(facilities[iFacility]);
     GDALRasterBand* facilityBand = facilityDataset->GetRasterBand(1);
     BYTE facilityNodata = facilityBand->GetNoDataValue();
-
     CPLAssert(facilityBand->GetRasterDataType() == GDT_Byte);
-    int xBlockSizeFacility, yBlockSizeFacility;
-    facilityBand->GetBlockSize(&xBlockSizeFacility, &yBlockSizeFacility);
-    CPLAssert(xBlockSizeFacility == xBlockSize);
-    CPLAssert(yBlockSizeFacility == yBlockSize);
-    CPLAssert(facilityDataset->GetRasterXSize() == xSize);
-    CPLAssert(facilityDataset->GetRasterYSize() == ySize);
+
+    // set blocks offsets
+    double targetProjection[6], facilityProjection[6];
+    targetDataset->GetGeoTransform(targetProjection);
+    facilityDataset->GetGeoTransform(facilityProjection);
+
+    double facilityMaxY = facilityProjection[3];
+    double demoMaxY = facilityProjection[3];
+    // double facilityYRes = facilityProjection[5];
+    double demoYRes = targetProjection[5];
+    // TODO: assert that both res are equal
+    int blocksRowOffset = std::abs(round( (facilityMaxY-demoMaxY)/(128 * demoYRes) ));
```

Comment:
It was an attempt to prevent errors that could arose from floating point math when dealing with regions and isochrones with similar bounds. The abs and round are interrelated in this case because floor takes a negative number to it's previous integer which is greater than the searched value.  We could assume we will never get negative numbers there and use floor without abs but I suggest at least adding an assertion of that fact. What do you think?